README
$Progeny$

This is the Componentized Linux Platform Development Kit (PDK).

    Copyright 2005 Progeny Linux Systems, Inc.

    PDK is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    PDK is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with PDK; if not, write to the Free Software Foundation,
    Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

See http://componentizedlinux.org/ for contact information and
documentation.

To install:
python setup.py install

vim:set ai et sw=4 ts=4 tw=75:


debian/changelog
pdk (1.0.0~alpha3) unstable; urgency=low

  * [pdk]
    - listmeta, listcomps: add new commands for pdk-mediagen
    - repogen: hide output of gpgv correctly

 -- Christopher Obbard <chris@64studio.com>  Tue, 26 Jun 2018 19:48:43 +0100

pdk (1.0.0~alpha2) unstable; urgency=low

  * [debian]
    - add Daniel James <daniel@64studio.com> as uploader
    - replace git-core dependency with git in line with Debian Buster
    - add gnupg dependency

  * [pdk]
    - repogen: only generate repository if GPG private key exists
    - repogen: export GPG public key and keyring into repository
    - repogen: hide output from gpg when signing
    - mediagen: refactor command to call command in package pdk-mediagen
    - pull: use correct merge command
    - components: allow comments inside meta tag

  * [mediagen]
    - moved mediagen into its own package: pdk-mediagen

 -- Christopher Obbard <chris@64studio.com>  Sat, 16 Jun 2018 16:01:26 +0100

pdk (1.0.0~alpha1) unstable; urgency=low
  
  * [debian]
    - update packaging in line with Jessie
    - set myself as maintainer
  * [vc]
    - keep track of etc/channels.xml file
    - use local ignore files, if present
    - change format of log command to be more useful
    - allow pulling from local projects
    - disable framing pulling/pushing as it is broken
  * [pdk]
    - allow referencing deb files inside workspace
    - allow overriding package priority from meta
    - fix abstract command crashing when no packages are passed
    - allow multiple selections for abstact command
    - add stronger hashing to apt repo generation in wheezy
  * [mediagen]
    - initial commits of next-gen media generator command

 -- Christopher Obbard <chris@64studio.com>  Mon, 06 Mar 2017 21:06:52 +0000

pdk (0.9.9) unstable; urgency=low

  * [debian]
    - format package description to have lines shortern than 80 chars
    - move Homepage from description to the source package stanza
    - drop picax scripts from the binary package, they are now
      merged with the main pdk script
    - bump standards version
    - first upload to Debian (Closes: #508112)

 -- Free Ekanayaka <freee@debian.org>  Mon, 08 Dec 2008 01:29:44 +0100

pdk (0.9.8) 64studio; urgency=low

  * [debian]
    - generate changelog when building master images
    - update package description
    - add Makefile.project example
  * [closure]
    - use UTF-8 strings to deal with smart generated error messages
  * [mediagen]
    - set the architecture from global_conf in /etc/apt-info/apt.conf
    - initialize config['installer_options'] to avoid breaking mediagen -a
    - remove apt lock file after building
    - remove temporary repositories after building the media
    - set the iso file name with the output-dest option

 -- Free Ekanayaka <freee@debian.org>  Mon, 01 Dec 2008 07:52:37 +0100

pdk (0.9.7) 64studio; urgency=low

  * Include isolinux.bin, as the one shipped by Lenny breaks pdk
  * Add inst-timeout option to set isolinux boot timeout

 -- Free Ekanayaka <freee@debian.org>  Wed, 15 Oct 2008 12:33:14 +0200

pdk (0.9.6) 64studio; urgency=low

  * Update Homepage in debian/control
  * Remove README, it is a duplicate of debian/copyright

 -- Free Ekanayaka <freee@debian.org>  Fri, 03 Oct 2008 11:51:41 +0100

pdk (0.9.5) 64studio; urgency=low

  * Add .gpg.conf and set debian-tag to %(versio)s
  * Clean apt temp directory after running the mediagen command

 -- Free Ekanayaka <freee@debian.org>  Sat, 27 Sep 2008 09:54:53 +0100

pdk (0.9.4) 64studio; urgency=low

  * Applied patches
  * Ported to python 2.5
  * Depends on rpm
  * Set myself as maintainer
  * Run git-init when creating a new workspace
  * Added missing build-dependences
  * Depend on smartpm-core instead of smartpm
  * Renamed mkisofs in genisoimage
  * Use media-label instead of cd-label to write the .disk/info file
  * Cleanup previous changelog entries
  * Added support for bash-like wildcard '*' in the --exclude option of
    the abstract command
  * Strip blanks from exclude patterns
  * New --packages option for the abstract command
  * Let the closure command write to an existing component through the -
    -out-file option
  * Added --label option to the abstract command, for setting the id of
    the generated component
  * Change working directory to the current workspace before running
    commit commands
  * Some more comments in run_closure
  * Removed smartpm hack, bug reported upstream, using smartpm << 0.51
    from now on
  * Depend on smartpm << 0.51 as higher versions break the closure
    command
  * Reverted pdk-deb channel priority to 1000, as setting it to 100
    breaks smart's resolver
  * Remove atest.tar.cdbs-config_list when cleaning
  * Run git-reset after the pull command
  * Add --components option to the abstract command
  * Add *.pyc and debian build files to .gitignore
  * Disable audit, push, update, dumplinks and migrate commands. They
    are not stable yet

 -- Free Ekanayaka <freee@debian.org>  Sat, 27 Sep 2008 09:41:49 +0200

pdk (0.9.3) unstable; urgency=low

  * Use newer Debian policy for building.
  * Use versioned dep for git-core.
  * Update to use newer pylint.
  * Fix push command.

 -- Darrin Thompson <darrint@progeny.com>  Thu,  5 Oct 2006 16:44:02 -0400

pdk (0.9.2) unstable; urgency=low

  * Fix Debian control file to assign package to "devel" section, not "dev".

 -- Branden Robinson <branden@progeny.com>  Thu,  3 Aug 2006 13:06:45 -0400

pdk (0.9.1) unstable; urgency=low

  * In conditional includes, s/narrow/limit/, s/mask/exclude/.

 -- Darrin Thompson <darrint@progeny.com>  Fri, 21 Jul 2006 09:46:32 -0400

pdk (0.9.0) unstable; urgency=low

  * Bump version to 0.9.0, since we are _way_ past 0.0.x.
  * Add mv command.
  * Add aliases create workspace and rm.
  * Clean up diagnostics.
  * Add rpm-md channels.

 -- Darrin Thompson <darrint@progeny.com>  Fri, 14 Jul 2006 12:19:03 -0400

pdk (0.0.37) unstable; urgency=low

  * Fill out pdk dependencies in pdk.spec.

pdk/audit.py
# $Progeny$
#
#   Copyright 2005 Progeny Linux Systems, Inc.
#
#   This file is part of PDK.
#
#   PDK is free software; you can redistribute it and/or modify it
#   under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   PDK is distributed in the hope that it will be useful, but WITHOUT
#   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
#   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
#   License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with PDK; if not, write to the Free Software Foundation,
#   Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

"""
audit

Load the component(s), and verify that all of its parts exist in the
cache Part of the PDK suite
"""
__revision__ = '$Progeny$'

from sets import Set
import sha
import md5
import pdk.workspace as workspace
from pdk.component import ComponentDescriptor
from pdk.command_base import make_invokable
import pdk.log as log
from pdk.exceptions import IntegrityFault
audit_logger = log.get_logger()


def audit(args):
    """\\fB%prog\\fP
.PP
Load the component,
and verify that it and it's parts are well-formed.
    """
    ##specialization code starts here

    def note_problem(fact, prediction, prediction_basis,
                     evidence, evidence_basis):
        """A mismatch handler for Arbiter

        Handle the case of a prediction that doesn't pan out
        by reporting the mismatch to stdout, in a format
        suitable for: cut -d'|'
        """
        note_problem.called = True
        fields = fact.get_problem_description(prediction, prediction_basis,
                                         evidence, evidence_basis)
        result = string_together(fields, '|')
        print result

    note_problem.called = False

    my_cache = workspace.current_workspace().cache
    arbiter = Arbiter(note_problem)

    for component_name in args.args:
        def _note_blob_id(blob_id):
            """Make common predictions and warrants for blob_id.

            Predict that a blob_id will be in cache
            Warrant the blob_id will be needed by the component.
            """
            arbiter.predict(InCache(blob_id), True, component_name)

        # Get the set of packages in the component
        descriptor = ComponentDescriptor(component_name)
        component = descriptor.load(my_cache)

        # predict expected blob_ids and headers
        for package in component.iter_packages():
            _note_blob_id(package.blob_id)
            arbiter.predict(InCache(package.blob_id + '.header'),
                            True, component_name)
            for package_tuple in package.extra_files:
                blob_id = package_tuple[0]
                _note_blob_id(blob_id)

        # predict upcoming source packages
        for package in component.iter_packages():
            if package.role == 'binary':
                fact = HasSource(package.format, package.pdk.sp_name,
                                 package.pdk.sp_version.full_version)
                arbiter.predict(fact, True, component_name)

        # warrant source packages found
        for package in component.iter_packages():
            if package.role == 'source':
                fact = HasSource(package.format, package.name,
                                 package.version.full_version)
                arbiter.warrant(fact, True, component_name)

    # predict upcoming cache checksums
    # pre-warrant found inodes
    # note inode -> filename relationships for later
    found_by_inode = {}
    for blob_id in my_cache:
        inode = my_cache.get_inode(blob_id)
        if blob_id.endswith('.header'):
            arbiter.warrant(InCache(blob_id), True, 'cache')
            continue

        entry = found_by_inode.setdefault(inode, Set([]))
        entry.add(blob_id)

        # ? Won't this create repeated predictions?
        for blob_id in found_by_inode[inode]:
            arbiter.predict(ChecksumMatches(blob_id), blob_id, 'cache')

    for inode, blob_ids in found_by_inode.iteritems():
        for blob_id in blob_ids:
            arbiter.warrant(InCache(blob_id), True, 'cache')

        # warrant cache checksums
        one_id = tuple(blob_ids)[0]
        handle = open(my_cache.file_path(one_id))
        sha1_digest = sha.new()
        md5_digest = md5.new()
        read_size = 8196
        while 1:
            block = handle.read(read_size)
            if not block:
                break
            sha1_digest.update(block)
            md5_digest.update(block)
        handle.close()

        prefixes = []
        for blob_id in blob_ids:
            if blob_id.startswith('sha-1'):
                prefixes.append('sha-1')
                arbiter.warrant(ChecksumMatches(blob_id),
                                'sha-1:' + sha1_digest.hexdigest(),
                                'cache')
            elif blob_id.startswith('md5'):
                prefixes.append('md5')
                arbiter.warrant(ChecksumMatches(blob_id),
                                'md5:' + md5_digest.hexdigest(), 'cache')
            else:
                # note unknown prefixes
                arbiter.note_problem(
                    blob_id
                    , ('md5:', 'sha-1:')
                    , 'unknown prefix'
                    )
        prefixes.sort()
        if prefixes != ['md5', 'sha-1']:
            digests = (md5_digest.hexdigest(), sha1_digest.hexdigest())
            arbiter.note_problem(tuple(blob_ids), digests,
                                 'not hard linked properly')

    arbiter.note_leftovers()

    if note_problem.called:
        raise IntegrityFault, "Audit detected fault(s)"

audit = make_invokable(audit)

def string_together(fields, separator):
    """Stringify fields and join them with separator.

    if not bool(field): stringified field becomes ''.
    """
    string_fields = [ f and str(f) or '' for f in fields ]
    joined_string = separator.join(string_fields)
    return joined_string

class Arbiter(object):
    """Match predictions with warrants.

    Predictions do not neccessarily have to precede warrants.

    Warranted facts which are never predicted are ignored.

    mismatch_handler is called as possible when a mismatch is found.

    Call note_leftovers after adding all predictions and warrants to
    mop up any dangling predictions.

    Call note_problem to note problems which do not require a
    prediction/warrant pair.
    """
    __slots__ = ('prediction_index', 'warrant_index', 'mismatch_handler')

    def __init__(self, mismatch_handler):
        self.mismatch_handler = mismatch_handler
        self.prediction_index = {}
        self.warrant_index = {}

    def predict(self, fact, prediction, prediction_basis):
        """Predict that evidence will assert a fact."""
        if fact in self.warrant_index:
            evidence, warrant_basis = self.warrant_index.pop(fact)
            if prediction != evidence:
                self.mismatch_handler(fact, prediction, prediction_basis,
                                 evidence, warrant_basis)
        else:
            self.prediction_index[fact] = (prediction, prediction_basis)

    def warrant(self, fact, evidence, warrant_basis):
        """Warrant that evidence has been found to establish a fact."""
        if fact in self.prediction_index:
            prediction, prediction_basis = self.prediction_index.pop(fact)
            if prediction != evidence:
                self.mismatch_handler(fact, prediction, prediction_basis,
                                 evidence, warrant_basis)
        else:
            self.warrant_index[fact] = (evidence, warrant_basis)

    def note_leftovers(self):
        """Note any dangling predictions."""
        for fact in dict(self.prediction_index):
            self.warrant(fact, None, 'no warrant')

    def note_problem(self, subject, prediction, message):
        """Directly note a problem; No prediction/warrant pair is needed.
        """
        fact = AlwaysFail(message)

        self.mismatch_handler(fact, prediction, None, None, subject)

class FactType(object):
    """Base class for fact types.

    Requires the use of a __slots__ field.

    Automatically makes derived types hashable, comparable, and iterable
    on the fields in their slot list.
    """
    # Our facts are made to work with the 'note_problem'
    def __init__(self, *args):
        arity = len(self.__slots__) + 1
        args_len = len(args) + 1
        if args_len != arity:
            raise TypeError, '__init__() takes %d arguments %d given.' \
                  % (arity, args_len)
        for name, arg in zip(self.__slots__, args):
            setattr(self, name, arg)

    def __iter__(self):
        return iter([ getattr(self, x) for x in self.__slots__ ])

    def __cmp__(self, other):
        return cmp(type(self), type(other)) or cmp(list(self), list(other))

    def __hash__(self):
        return hash((type(self), tuple(self)))

class InCache(FactType):
    """The blob_id is expected to be in the cache."""
    __slots__ = 'blob_id',
    blob_id = None

    def get_problem_description(self, *args):
        """Return data suitable for Arbiter.mismatch_handler."""
        #prediction, prediction_basis, evidence, evidence_basis
        prediction_basis = args[1]
        return (self.blob_id, prediction_basis, 'missing from cache')

class InodeNeeded(FactType):
    """The inode is needed by some component."""
    __slots__ = 'inode',
    inode = None

    def get_problem_description(self, *args):
        """Return data suitable for Arbiter.mismatch_handler."""
        #prediction, prediction_basis, evidence, evidence_basis
        prediction_basis = args[1]
        return (prediction_basis, '', 'not needed by components')

class HasSource(FactType):
    """The given fields represent a source package in a component."""
    __slots__ = 'format', 'name', 'version'
    format = None
    name = None
    version = None

    def get_problem_description(self, *dummy):
        """Return data suitable for Arbiter.mismatch_handler."""
        subject = string_together([self.format, self.name, self.version],
                                  ',')
        return (subject, '', 'missing source')

class ChecksumMatches(FactType):
    """The the contents of the blob_id match it's checksum."""
    __slots__ = 'blob_id',
    blob_id = None

    def get_problem_description(self, *args):
        """Return data suitable for Arbiter.mismatch_handler."""
        # args= prediction, dummy, evidence, ...
        prediction, evidence = args[0], args[2]
        return (prediction, evidence, 'checksum mismatch')

class AlwaysFail(FactType):
    """Fail directly. Used by Arbiter.note_problem."""
    __slots__ = 'message',
    message = None

    def get_problem_description(self, *args):
        """Return data suitable for Arbiter.mismatch_handler."""
        prediction, evidence_basis = args[0], args[3]
        return (evidence_basis, prediction, self.message)

# vim:set ai et sw=4 ts=4 tw=75:
